compiler/optimizer/ValuePropagationCommon.cpp

```diff
@@ -4090,17 +4146,31 @@ void OMR::ValuePropagation::transformArrayCloneCall(TR::TreeTop *callTree, OMR::
    newArray->setIsNonNull(true);
 
    int32_t elementSize = TR::Compiler->om.getSizeOfArrayElement(newArray);
-   TR::Node *lengthInBytes = comp()->target().is64Bit() ?
-      TR::Node::create(callNode, TR::lmul, 2, TR::Node::create(callNode, TR::i2l, 1, lenNode), TR::Node::lconst(lenNode, elementSize)) :
-      TR::Node::create(callNode, TR::imul, 2, lenNode, TR::Node::iconst(lenNode, elementSize));
 
+   TR::Node *lengthInBytes;
+   TR::Node *srcStart;
+   TR::Node *destStart;
+#if defined(OMR_GC_SPARSE_HEAP_ALLOCATION)
+   if (TR::Compiler->om.isOffHeapAllocationEnabled())
+      {
+      lengthInBytes = TR::TransformUtil::generateArrayOffsetTrees(comp(), lenNode, NULL, elementSize, false);
+      srcStart = TR::TransformUtil::generateArrayStartTrees(comp(), objNode);
+      destStart = TR::TransformUtil::generateArrayStartTrees(comp(), newArray);
+      }
+   else
+#endif /* OMR_GC_SPARSE_HEAP_ALLOCATION */
+      {
+      lengthInBytes = comp()->target().is64Bit() ?
+         TR::Node::create(callNode, TR::lmul, 2, TR::Node::create(callNode, TR::i2l, 1, lenNode), TR::Node::lconst(lenNode, elementSize)) :
+         TR::Node::create(callNode, TR::imul, 2, lenNode, TR::Node::iconst(lenNode, elementSize));
+      srcStart = comp()->target().is64Bit() ?
+         TR::Node::create(callNode, TR::aladd, 2, objNode, TR::Node::lconst(objNode, TR::Compiler->om.contiguousArrayHeaderSizeInBytes())) :
+         TR::Node::create(callNode, TR::aiadd, 2, objNode, TR::Node::iconst(objNode, TR::Compiler->om.contiguousArrayHeaderSizeInBytes()));
+      destStart = comp()->target().is64Bit() ?
+         TR::Node::create(callNode, TR::aladd, 2, newArray, TR::Node::lconst(newArray, TR::Compiler->om.contiguousArrayHeaderSizeInBytes())) :
+         TR::Node::create(callNode, TR::aiadd, 2, newArray, TR::Node::iconst(newArray, TR::Compiler->om.contiguousArrayHeaderSizeInBytes()));
+      }
```

Comment:
The transformation doesn't do anything if arraylets are enabled so we should be ok using `TR::TransformUtil::generateArrayStartTrees()`, no?

[1] [compiler/optimizer/ValuePropagationCommon.cpp#L4001](https://github.com/eclipse/omr/blob/1d182ba9273967af891845e0926c4feed50c91d2/compiler/optimizer/ValuePropagationCommon.cpp#L4057) "Arraylets check"

Reply:
Will test again, with current implementation of `TR::Compiler->om.canGenerateArraylets()` it should work yes.

Reply:
Modified code as mentioned